### ISSUE TITLE
implement stronger refinement technique

### ISSUE THREAD

**Issue by proteneer** [proteneer/bgl#11]
currently, we refine the marcs matrix only by the rows, however, a stronger (and what should've been obvious condition), is to also refine by columns based on the matches in the second mol.

**Comment by proteneer** [comment]
note that this doesn't have a huge impact on # of nodes visited for hif2a:

master:
```
====[NODES VISITED 808 | CORE_SIZE 34 | NUM_EDGES 34 | time taken: 0.015382051467895508 | time out? False]=====
====[NODES VISITED 107 | CORE_SIZE 37 | NUM_EDGES 38 | time taken: 0.0022897720336914062 | time out? False]=====
====[NODES VISITED 110 | CORE_SIZE 37 | NUM_EDGES 39 | time taken: 0.002119779586791992 | time out? False]=====
====[NODES VISITED 1137 | CORE_SIZE 31 | NUM_EDGES 31 | time taken: 0.02042388916015625 | time out? False]=====
====[NODES VISITED 99 | CORE_SIZE 34 | NUM_EDGES 34 | time taken: 0.0019352436065673828 | time out? False]=====
====[NODES VISITED 107 | CORE_SIZE 34 | NUM_EDGES 35 | time taken: 0.0020329952239990234 | time out? False]=====
====[NODES VISITED 73 | CORE_SIZE 31 | NUM_EDGES 31 | time taken: 0.0014319419860839844 | time out? False]=====
====[NODES VISITED 158 | CORE_SIZE 37 | NUM_EDGES 39 | time taken: 0.002938985824584961 | time out? False]=====
====[NODES VISITED 119 | CORE_SIZE 31 | NUM_EDGES 31 | time taken: 0.002276897430419922 | time out? False]=====
====[NODES VISITED 145 | CORE_SIZE 31 | NUM_EDGES 32 | time taken: 0.0026760101318359375 | time out? False]=====
```

with stronger refinement:
```
====[NODES VISITED 757 | CORE_SIZE 34 | NUM_EDGES 34 | time taken: 0.016103267669677734 | time out? False]=====
====[NODES VISITED 98 | CORE_SIZE 37 | NUM_EDGES 38 | time taken: 0.0022246837615966797 | time out? False]=====
====[NODES VISITED 101 | CORE_SIZE 37 | NUM_EDGES 39 | time taken: 0.0021300315856933594 | time out? False]=====
====[NODES VISITED 1016 | CORE_SIZE 31 | NUM_EDGES 31 | time taken: 0.020481109619140625 | time out? False]=====
====[NODES VISITED 84 | CORE_SIZE 34 | NUM_EDGES 34 | time taken: 0.0018219947814941406 | time out? False]=====
====[NODES VISITED 107 | CORE_SIZE 34 | NUM_EDGES 35 | time taken: 0.0021948814392089844 | time out? False]=====
====[NODES VISITED 73 | CORE_SIZE 31 | NUM_EDGES 31 | time taken: 0.0015747547149658203 | time out? False]=====
====[NODES VISITED 158 | CORE_SIZE 37 | NUM_EDGES 39 | time taken: 0.0031919479370117188 | time out? False]=====
====[NODES VISITED 119 | CORE_SIZE 31 | NUM_EDGES 31 | time taken: 0.0025162696838378906 | time out? False]=====
====[NODES VISITED 145 | CORE_SIZE 31 | NUM_EDGES 32 | time taken: 0.0029931068420410156 | time out? False]=====
```

**Comment by proteneer** [comment]
However, for harder systems this seems to be a bit more dramatic (100x):

before:
```
====[NODES VISITED 4920904 | CORE_SIZE 62 | NUM_EDGES 58 | time taken: 73.29053497314453 | time out? False]=====
```
after:
```
====[NODES VISITED 40291 | CORE_SIZE 62 | NUM_EDGES 58 | time taken: 0.728355884552002 | time out? False]=====
```

**Comment by proteneer** [comment]
implemented in #12
